compiler/rustc_infer/src/infer/error_reporting/mod.rs

```diff
@@ -1917,6 +1917,14 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
                     );
                     if !is_simple_error || terr.must_include_note() {
                         diag.note_expected_found(&expected_label, expected, &found_label, found);
+
+                        if let Some(expected_type_found) = exp_found {
+                            if let ty::Closure(_, args) = expected_type_found.found.kind() {
```

Comment:
Does it make sense to add something similar for `async` closures and friends?

Reply:
I don't think it's necessary for now.